ark/utils/metacluster_remap_gui/metaclustergui.py

```diff
@@ -0,0 +1,463 @@
+        self.current_metacluster = widgets.Dropdown(
+            value=self.mcd.metaclusters.index[0],
+            options=list(zip(self.mcd.metacluster_displaynames, self.mcd.metaclusters.index)),
+            description='MetaCluster:',
+            )
+        self.current_metacluster.observe(
+            self.update_current_metacluster_handler, type="change", names="value")
+
+        self.current_metacluster_displayname = widgets.Text(
+            value=self.mcd.get_metacluster_displayname(self.current_metacluster.value),
+            placeholder='Metacluster Displayname',
+            description='Edit Name:',
+            disabled=False,
+            )
+        self.current_metacluster_displayname.observe(
+            self.update_current_metacluster_displayname,
+            type="change",
+            names="value")
+
+        # group widgets to look nice
+        self.metacluster_info = widgets.VBox([
+            self.current_metacluster,
+            self.current_metacluster_displayname])
+        self.tools = widgets.HBox([
+            self.zscore_clamp_slider,
+            self.clear_selection_button,
+            self.new_metacluster_button,
+            ])
+        self.toolbar = widgets.HBox([
+            self.tools,
+            self.metacluster_info,
+            ])
+        self.toolbar.layout.justify_content = 'space-between'
+        display(self.toolbar)
+
+    def move_dendro_labels(self, ax, dendrosplit_ratio=1.8):
+        """Overlay axis labels directly onto a scipy dendrogram
+
+        Final image will use the ratio 1:dendrosplit_ratio
+        for tree_region:labels_region
+        """
+        def add_room_for_labels():
+            ax.set_axisbelow(False)
+            xlim = ax.get_xlim()
+            ax.set_xlim((xlim[0], -(xlim[0]*dendrosplit_ratio)))
+
+        def stretch_dendro_leaves():
+            for c in ax.collections:
+                for path in c.get_paths():
+                    for v in path.vertices:
+                        if v[0] == 0:
+                            v[0] = ax.get_xlim()[1]
+
+        def get_ax_width_points(ax):
+            """points = 1/72 in"""
+            bbox = ax.get_window_extent().transformed(ax.figure.dpi_scale_trans.inverted())
+            return bbox.width * 72
+
+        def move_ax_labels():
+            dr = dendrosplit_ratio
+            width = get_ax_width_points(ax)
+            dedent = -(width * dr / (1 + dr))
+            ax.yaxis.set_tick_params(which='both', pad=dedent)
+
+        def restyle_ax_labels():
+            for lb in ax.get_yticklabels():
+                lb.set_path_effects([
+                    path_effects.Stroke(linewidth=4, foreground='white'),
+                    path_effects.Normal(),
+                    ])
+                lb.set_family('monospace')
+                lb.set_zorder(4)
+
+        add_room_for_labels()
+        stretch_dendro_leaves()
+        move_ax_labels()
+        restyle_ax_labels()
```

Comment:
Any reason we need to do it this way, and we can't do something instead like:

```
# add room for axes labels
ax.set_axisbelow(False)
xlim = ax.get_xlim()
ax.set_xlim((xlim[0], -(xlim[0]*dendrosplit_ratio)))

# stretch dendro leaves
# code in stretch_dentro_leaves

...
```

Reply:
Could be done either way, I look into it, might have started out longer than it ended up.

Two clean code heuristics explain why I did it this way.

1. Single level of abstraction
2. Prefer self-documenting function names over comments

## Single level of abstraction
http://principles-wiki.net/principles:single_level_of_abstraction
All statements of a method should belong to the same level of abstraction. If there is a statement which belongs to a lower level of abstraction, it should go to a private method which comprises statements on this level. Doing so will result in smaller methods.

## Prefer self-documenting function names over comments
https://levelup.gitconnected.com/instead-of-a-comment-use-a-well-named-function-448eabb45607

Reply:
Ok after sleeping on it I realize you probably asked because I used nested defs. It's to decrease the scope of the def itself (e.g. private methods). Basically its a way of getting benefits of a static helper object with private methods but without overcomplicating architecture.

In languages without nested defs I would have created a separate class with the nested def put in as private methods.